engine/api/v2_workflow_run_engine.go

```diff
@@ -0,0 +1,384 @@
+				log.ErrorWithStackTrace(ctx, err)
+			}
+		}
+	}
+}
+
+func (api *API) V2WorkflowRunEngineDequeue(ctx context.Context) error {
+	for {
+		var wrEnqueue sdk.V2WorkflowRunEnqueue
+		if err := api.Cache.DequeueWithContext(ctx, workflow_v2.WorkflowEngineKey, 250*time.Millisecond, &wrEnqueue); err != nil {
+			log.Error(ctx, "V2WorkflowRunEngine > DequeueWithContext err: %v", err)
+			continue
+		}
+		if err := api.workflowRunV2Trigger(ctx, wrEnqueue); err != nil {
+			log.ErrorWithStackTrace(ctx, err)
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+	}
+}
+
+// TODO Manage stage
+// TODO Manage job sub number
+// TODO manage git context
+// TODO manage vars context
+func (api *API) workflowRunV2Trigger(ctx context.Context, wrEnqueue sdk.V2WorkflowRunEnqueue) error {
+	_, next := telemetry.Span(ctx, "api.workflowRunV2Trigger.lock")
+	lockKey := cache.Key("api:workflow:engine", wrEnqueue.RunID)
+	b, err := api.Cache.Lock(lockKey, 5*time.Minute, 0, 1)
+	if err != nil {
+		next()
+		return err
+	}
+	if !b {
+		log.Debug(ctx, "api.workflowRunV2Trigger> run %d is locked in cache", wrEnqueue.RunID)
+		// re-enqueue workflow-run
+		if err := api.Cache.Enqueue(workflow_v2.WorkflowEngineKey, wrEnqueue); err != nil {
+			next()
+			return err
+		}
+		next()
+		return nil
+	}
+	next()
+	defer func() {
+		_ = api.Cache.Unlock(lockKey)
+	}()
+
+	// Load run by id
+	_, next = telemetry.Span(ctx, "workflowRunV2Trigger.LoadRunByID")
+	run, err := workflow_v2.LoadRunByID(ctx, api.mustDB(), wrEnqueue.RunID)
+	if sdk.ErrorIs(err, sdk.ErrNotFound) {
+		next()
+		return nil
+	}
+	if err != nil {
+		next()
+		return sdk.WrapError(err, "unable to load workflow run %s", wrEnqueue.RunID)
+	}
+	next()
+
+	if sdk.StatusIsTerminated(run.Status) && len(wrEnqueue.Jobs) == 0 {
+		log.Debug(ctx, "workflow run already on a final state")
+		return nil
+	}
+
+	u, err := user.LoadByID(ctx, api.mustDB(), wrEnqueue.UserID)
+	if err != nil {
+		return err
+	}
+
+	jobsToQueue, runMsgs, currentJobRunStatus, errRetrieveJobs := retrieveJobToQueue(ctx, api.mustDB(), run, wrEnqueue, u, api.Config.Workflow.JobDefaultRegion)
+	log.Debug(ctx, "workflowRunV2Trigger> jobs to queue: %+v", jobsToQueue)
+
+	tx, errTx := api.mustDB().Begin()
+	if errTx != nil {
+		return sdk.WithStack(err)
+	}
+	defer tx.Rollback() // nolint
+
+	if errRetrieveJobs != nil {
+		for i := range runMsgs {
+			if err := workflow_v2.InsertRunInfo(ctx, tx, &runMsgs[i]); err != nil {
+				return err
+			}
+			run.Status = sdk.StatusFail
+			if err := workflow_v2.UpdateRun(ctx, tx, run); err != nil {
+				return err
+			}
+			if err := tx.Commit(); err != nil {
+				return err
+			}
+		}
+		return errRetrieveJobs
+	}
+
+	// Enqueue JOB
+	for jobID, jobDef := range jobsToQueue {
+		runJob := sdk.V2WorkflowRunJob{
+			WorkflowRunID: run.ID,
+			Status:        sdk.StatusWaiting,
+			JobID:         jobID,
+			Job:           jobDef,
+			UserID:        wrEnqueue.UserID,
```

Comment:
Here we have something to do regarding the user.
For run triggered by git commits, we have to handle the user that pushed the commit

Reply:
Aggreed. But this will not be here, but on handler side when we will develop hook manangement